File: tnzhou/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == NULL)
            return true;
        return validateBSTHelper(root->left, (long long) -2147483648 - 1, root->val) &&
            validateBSTHelper(root->right, root->val, (long long) 2147483647 + 1);
    }
    
    bool validateBSTHelper(TreeNode* node, long long down_limit, long long up_limit) {
        if (node == NULL)
            return true;
        if (node->val <= down_limit || node->val >= up_limit)
            return false;
        return validateBSTHelper(node->left, down_limit, node->val) &&
            validateBSTHelper(node->right, node->val, up_limit);
    }
};

File: zemiao/98.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        valid = [True]
        
        def getInOrder(root):
          if not root:
            return []
          res = [root.val, root.val]
          if root.left:
            left = getInOrder(root.left)
            if valid[0] and left[-1]<root.val:
              res = [left[0],res[1]]
            else:
              valid[0] = False
              return
          if root.right:
            right = getInOrder(root.right)
            if valid[0] and right[0]>root.val:
              res = [res[0], right[-1]]
            else:
              valid[0] = False
              return
          return res
        
        getInOrder(root)
        return valid[0]
        
        